lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,407 @@
+		return -EINVAL;
+	}
+
+	LOG_DBG("Sample rate converter initialized. Input sample rate: %d, Output sample rate: %d, "
+		"conversion ratio: %d, filter type: %d",
+		ctx->input_sample_rate, ctx->output_sample_rate, ctx->conversion_ratio,
+		ctx->filter_type);
+	return 0;
+}
+
+int sample_rate_converter_open(struct sample_rate_converter_ctx *ctx)
+{
+	if (ctx == NULL) {
+		LOG_ERR("Context cannot be NULL");
+		return -EINVAL;
+	}
+
+	memset(ctx, 0, sizeof(struct sample_rate_converter_ctx));
+
+	return 0;
+}
+
+int sample_rate_converter_process(struct sample_rate_converter_ctx *ctx,
+				  enum sample_rate_converter_filter filter, void *input,
+				  size_t input_size, uint32_t input_sample_rate, void *output,
+				  size_t output_size, size_t *output_written,
+				  uint32_t output_sample_rate)
+{
+	int ret;
+	uint8_t *read_ptr;
+	uint8_t *write_ptr;
+	size_t samples_to_process;
+
+	uint8_t internal_input_buf[SAMPLE_RATE_CONVERTER_INTERNAL_INPUT_BUF_SIZE];
+	uint8_t internal_output_buf[SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE];
+
+#if CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_16
+	size_t bytes_per_sample = sizeof(uint16_t);
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+	size_t bytes_per_sample = sizeof(uint32_t);
+#endif
+
+	if (input_size % bytes_per_sample != 0) {
+		LOG_ERR("Size of input is not a byte multiple");
+		return -EINVAL;
+	}
+
+	size_t samples_in = input_size / bytes_per_sample;
+
+	if (samples_in > CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE) {
+		LOG_ERR("Too many samples given as input");
+		return -EINVAL;
+	}
+
+	if ((ctx == NULL) || (input == NULL) || (output == NULL) || (output_written == NULL)) {
+		LOG_ERR("Null pointer received");
+		return -EINVAL;
+	}
+
+	if ((ctx->input_sample_rate != input_sample_rate) ||
+	    (ctx->output_sample_rate != output_sample_rate) || (ctx->filter_type != filter)) {
+		LOG_DBG("State has changed, re-initializing filter");
+		ret = sample_rate_converter_reconfigure(ctx, input_sample_rate, output_sample_rate,
+							filter);
+		if (ret) {
+			LOG_ERR("Failed to initialize converter (%d)", ret);
+			return ret;
+		}
+	}
+
+	if ((ctx->conversion_direction == CONVERSION_DIR_DOWN) &&
+	    (samples_in < ctx->conversion_ratio)) {
+		LOG_ERR("Number of samples in can not be less than the conversion ratio (%d) when "
+			"downsampling",
+			ctx->conversion_ratio);
+		return -EINVAL;
+	}
+
+	if (ctx->conversion_direction == CONVERSION_DIR_UP) {
+		*output_written = input_size * ctx->conversion_ratio;
+	} else {
+		*output_written = input_size / ctx->conversion_ratio;
+	}
+
+	if (*output_written > output_size) {
+		LOG_ERR("Conversion process will produce more bytes than the output buffer can "
+			"hold");
+		return -EINVAL;
+	}
+
+	if (*output_written > SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE) {
+		LOG_ERR("Conversion process will produce more bytes than the internal output "
+			"buffer can hold");
+		return -EINVAL;
+	}
+
+	if (!conversion_needs_buffering(ctx, ctx->conversion_ratio)) {
```

Comment:
Can we switch around this if check, so we check for `conversion_needs_buffering` and not `!conversion_needs_buffering` ?